Make local test runs using paver run faster by not running migrations

douglashall opened the pull request:
These changes allow for making the running of unit tests significantly faster in a devstack development environment. This is accomplished by providing an option to prevent all migrations from running during test setup and creating database tables directly from the model definitions. Before these changes we were preventing most migrations from running, but the ones that ran still took up a significant amount of time and made local test development quite painful.

Test settings are split into three different files:

**(cms,lms)/envs/test_base.py**
Contains what was previously in test.py minus the original MIGRATION_MODULES hack used to speed up testing runs by preventing *most* migrations.

**(cms,lms)/envs/test.py**
Imports test_base.py and includes the original MIGRATION_MODULES hack.

**(cms,lms)/envs/test_without_migrations.py**
Imports test_base.py and includes a new MIGRATION_MODULES hack which prevents *all* migrations.

You can now include a --without_migrations option with the `paver test_system` command which will use the test_without_migrations settings file. By specifying this option you will see a ~15s decrease in the amount of time it takes to run tests.

douglashall commented:
@nedbat @benpatterson Can you take a look at this PR when you get a chance? Thanks.

benpatterson commented:
Could we get an update in testing.rst as part of this?

I'm still thinking through the "CLI convention" if you will. For example, should this be something else like `--local` instead? Also, I know we have some tables that need to be initialized with data...how does this approach tie into that?

benpatterson commented:
BTW I like where you're going; just thinking it through still. My other hesitation is running tests differently in a local env vs Jenkins. (B/c I want to keep the "why did this 'break' on Jenkins" confusions at a minimum.) But I think you're making that somewhat obvious...I'm not sure if `test_local` is the right name for it though...perhaps something more obvious like `on_fly_migrations` or something along those lines?

douglashall commented:
@benpatterson Cool. I added a `--without_migrations` flag to make the CLI clearer. I left the `--settings` flag for specifying django settings because this is something I would expect to be able to do when running commands related to a django app. I updated `testing.rst` calling out the `--without_migrations` flag.

This will not help the issue related to django apps that use data migrations for initial app data setup. We already skip migrations for some apps and ideally apps would use fixtures as opposed to data migrations for setting up initial app data.

The main goal of this PR is to make test development faster. This changeset will not affect the current CI test setup.

douglashall commented:
jenkins run lettuce

douglashall commented:
@jzoldak Could you take a look at this PR when you get a chance?

benpatterson commented:
@douglashall I have a couple of nits. The one I really do think you should change is the comment in the `test.py`s. Once that's good, :+1: from me.

nedbat commented on docs/en_us/internal/testing.rst:
```diff
@@ -196,6 +196,16 @@ To run a single test format the command like this.
 
     paver test_system -t lms/djangoapps/courseware/tests/tests.py:ActivateLoginTest.test_activate_login
 
+It is also helpful to be able to speed up tests by skipping Django migrations during test development. To make test
```
It'd be great to wrap these lines at 80 chars, though I know the rest of the file doesn't do it.

nedbat commented on cms/envs/test.py:
```diff
@@ -1,277 +1,10 @@
 
 # This hack disables migrations during tests. We want to create tables directly from the models for speed.
 # See https://groups.google.com/d/msg/django-developers/PWPj3etj3-U/kCl6pMsQYYoJ.
 MIGRATION_MODULES = {app: "app.migrations_not_used_in_tests" for app in INSTALLED_APPS}
```
This file should be the one *with* the migrations, right? But this comment seems to indicate that we'll avoid the migrations? And it was here before, so why do we have to change anything? I'm confused.

douglashall replied:
This method of skipping migrations does not seem to work for all django apps. The new method of doing this in `test_base.py` does.

nedbat replied:
But if the goal of this was to disable migrations, why not just disable them the right way here, and not add the complexity of two different ways to run tests?

nedbat replied:
That is, it looks like the original intent was to not use migrations during tests.  So let's fix this code to achieve that intent.  Or do we want to have a with-migrations and without-migrations choice?  In that case, this code should be deleted, since it leaves us in a confusing half-state where some apps run migrations and some don't.

douglashall replied:
I did not want to change the way we run tests during CI because I do not know enough about the original intentions of the old way of skipping migrations. This was a way of making a targeted change for local test development. I think ideally we should be running migrations during CI, but that seems like a bigger discussion that is outside the scope of the value being added here.

douglashall replied:
@benpatterson Do we need to skip migrations during CI test runs?

nedbat replied:
At the very very least, let's update this comment to reflect our current understanding.  But ideally, we'd get away from a half-migrated state.

douglashall replied:
@macdiesel @doctoryes Can you weigh in on this discussion?

nedbat commented on pavelib/tests.py:
```diff
@@ -27,9 +27,11 @@
     ("failed", "f", "Run only failed tests"),
     ("fail_fast", "x", "Fail suite on first failed test"),
     ("fasttest", "a", "Run without collectstatic"),
+    ("without_migrations", "m", "Run without migrations"),
```
Might want to add "(overrides --settings=)" to be clear about the relationship between them.

benpatterson commented:
Ignorant request on my part perhaps, but do unit tests work if we do the same on CI, as @nedbat is suggesting? I've been here awhile, but I also do not know, really, the original intent. Knowing if this method--universally adopted--would or would not change results would be good to know. If you have a few mins, could you create a separate WIP branch and run it on Jenkins @douglashall ?

douglashall commented:
@benpatterson Looks like full unit test suite will not run without further work when we turn off all migrations.

https://github.com/edx/edx-platform/pull/11108

I think what we really want though is to not skip ANY migrations when running unit tests during CI because doing so would move us further away from a production-like environment. Skipping migrations during test development works well to speed things up a bit, but I think we always want CI to run all migrations.

I would like @macdiesel or @doctoryes to weigh in on this though because the original code in test settings to skip some migrations most likely came about during the Django 1.8 work and I'd like to get the history on that.

Thanks all for helping work through this.

douglashall commented on lms/envs/test.py:
```diff
@@ -1,562 +1,10 @@
 
 # This hack disables migrations during tests. We want to create tables directly from the models for speed.
 # See https://groups.google.com/d/msg/django-developers/PWPj3etj3-U/kCl6pMsQYYoJ.
 MIGRATION_MODULES = {app: "app.migrations_not_used_in_tests" for app in INSTALLED_APPS}
```
@doctoryes Any idea who added this line originally? I believe it was incorporated with the Django 1.8 changes. Should we be skipping migrations at all during CI testing?

macdiesel replied:
@douglashall lets ping @symbolist on this change.

> MIGRATION_MODULES = {app: "app.migrations_not_used_in_tests" for app in INSTALLED_APPS}

symbolist commented:
@douglashall and others, here is my understanding of the current migrations setup in tests.

1. Migrations have been disabled in unit tests since the "[beginning](https://github.com/edx/edx-platform/blob/b3676cd760ac41037fc994faf157b6afed5df883/lms/envs/test.py#L23)". The only benefit I can think of this is speed.
2. Lettuce and bokchoy tests do run migrations. However, we cache the latest state of the database after certain intervals so all the migrations are not run, but only the ones added since the last time the database state was cached.

There are two other options here:
1. We run migrations but do not cache the database state like how it is happening currently in the other suites. But as the count of migrations grows much larger than the number of models, the units tests will become proportionally slower. This makes sense if running all the migrations from scratch is important.
2. We run migrations but cache the database state. So we only have to run the recent migrations.

Which option we pick depends on the answers to these questions:
1. Is is really important to run all migrations from the beginning during the unit tests?
2. If not, since the recent migrations are running as part of the the bokchoy tests, is it important to have them also run during unit tests?

benpatterson commented:
Responding to @symbolist, and perhaps taking it from a slightly different perspective:
* It would be great to test migrations from scratch. We were commonly in a state pre-django-1.8 where migrations from scratch basically did not work. We eventually found those problems downstream, but explicitly making sure migrations from scratch will work...would be extremely valuable.
* That said, I think it makes more sense to do that as an explicit test, rather than doing it as the foundation for the unit tests themselves.

symbolist commented:
One other thing to consider is that some migrations are used for seeding data. With our current setup for unit tests, this data is not added to the database. I suspect this is why some unit tests are failing in https://github.com/edx/edx-platform/pull/11108.

If we want this data available for unit tests then migrations+db caching would be the best solution.

benpatterson commented:
For the migrations that seed data, would the better (more Django-ey) solution be to use db fixtures instead? No idea how large a refactoring I'm suggesting (nor am I implying it be part of this particular PR), but just want to understand...

symbolist commented:
@benpatterson My understanding is that fixtures are targeted for tests, where a fresh database is created every time. If used when running an app, fixtures will get loaded on every migrate call and overwrite changes in the database.

https://docs.djangoproject.com/en/1.8/howto/initial-data/#automatically-loading-initial-data-fixtures
